### PR TITLE
Fixed bugs in safel

### PR DESCRIPTION
**Pull request by tiantianwdy**
Fixing issue #60  about safe serialization of EncryptedNumber.

* using reflection for default java serialization. 
* Add unit test for default java serialization.

### CODE REVIEW

File: src/test/java/com/n1analytics/paillier/PaillierEncryptedNumberTest.java [hunk 1]
```diff
@@ -1151,6 +1152,46 @@ public void testGetThisThingInSafe() {
...
+    @Test
+    public void testJavaSerialization () throws IOException, ClassNotFoundException {
+      EncryptedNumber unSafeEN = context.encrypt(context.encode(1.01, 1e-8));
+      assertFalse(unSafeEN.isSafe);
+      EncryptedNumber safeEN = unSafeEN.getSafeEncryptedNumber();
+      assertTrue(safeEN.isSafe);
+      assertNotEquals(unSafeEN.ciphertext, safeEN.ciphertext);
+
+      //test writing and reading safe en
+      String file = "./temp";
```
Comment (mhsiah) [inline src/test/java/com/n1analytics/paillier/PaillierEncryptedNumberTest.java]: You should use `File.createTempFile()` to create temporary files, and call `myfile.deleteOnExit()` to make sure it's removed.

Comment (tiantianwdy) [reply src/test/java/com/n1analytics/paillier/PaillierEncryptedNumberTest.java]: Fixed in latest commit. Feel free to let me know if there is further problems or questions:)

**Comment by wilko77** [comment]
### This naive way of serialisation is very wasteful.

Let's have a look at what you will actually serialise when serialising an EncryptedNumber (assume a 1024 bit key):
 - BigInteger ciphertext: ~256 bytes
 - int exponent: 4 bytes
 - boolean isSafe: 1 byte
 - PaillerContext context:
     - PaillierPublicKey publicKey:
         - BigInteger modulus: ~ 128 bytes
         - BigInteger modulusSquared: ~ 256 bytes
         - BigInteger generator: ~ 128 bytes
     - EncodingScheme encoding:
         - double log2Base: 8 bytes
         - boolean signed: 1 byte
         - int precision: 4 bytes
         - BigInteger maxEncoded: ~128 bytes
         - BigInteger minEncoded: ~128 bytes
         - BigInteger maxSignificand: ~128 bytes
         - BigInteger minSignificand: ~128 bytes
         - int base: 4 bytes

That's at least 1300 bytes... for each EncryptedNumber object. 

There is a lot of redundant information though:
 - a PaillierPublicKey can be fully described by `modulus`, that's a nice reduction from ~512 bytes to just ~128 bytes
 - once you got the public key, you just need `boolean signed`, `int precision`, and `int base` to create a PaillierContext with StandardEncodingScheme. That's 9 bytes vs. the naive ~530 bytes

I doubt the usefulness of the current proposition, as both, serialising to disk, and serialising to the network are heavily penalised with this ~300% blowup (actual transferred data vs. necessary data).

Another issue is, that once you want to serialise more than one number, serialising the context more than once is unnecessary.

### Object vs Data Serialisation
Serialising objects opens an attack vector where an attacker can inject malicious code into the serialised object. See for example https://www.slideshare.net/frohoff1/appseccali-2015-marshalling-pickles. 
I am not sure if this vulnerability applies here, the interested reader can find more info [here](https://github.com/GrrrDog/Java-Deserialization-Cheat-Sheet) or [here](https://www.owasp.org/index.php/Deserialization_Cheat_Sheet).

In my opinion, Javallier, as a security library, should adopt a more defensive approach and only explicitly allow data serialisation.

### Alternatives
The JavallierCLI contains code which (de-)serialises keys and encrypted numbers from/into JSON. That might be a good starting point.

**Comment by tiantianwdy** [comment]
With regard to Wilko's comment, I got one possible solution using reflection (still ugly), as below:

``` java
  private void writeObject(ObjectOutputStream out) throws IOException {
    EncryptedNumber en = isSafe? this : obfuscate();
    out.writeObject(en.ciphertext);
    out.writeInt(en.exponent);
    out.writeBoolean(en.isSafe);
    out.writeObject(en.context.getPublicKey().modulus);
    out.writeBoolean(en.context.getEncodingScheme().isSigned());
    out.writeInt(en.context.getEncodingScheme().getPrecision());
  }

  private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException, NoSuchFieldException, IllegalAccessException {
    Class c = EncryptedNumber.class;
    Field contextF = c.getDeclaredField("context");
    Field ciphertextF = c.getDeclaredField("ciphertext");
    Field exponentF = c.getDeclaredField("exponent");
    Field isSafeF = c.getDeclaredField("isSafe");
    contextF.setAccessible(true);
    ciphertextF.setAccessible(true);
    exponentF.setAccessible(true);
    isSafeF.setAccessible(true);
    BigInteger _ciphertext = (BigInteger) in.readObject();
    int _exponent = in.readInt();
    boolean _isSafe = in.readBoolean();
    BigInteger modulus = (BigInteger) in.readObject();
    boolean isSigned = in.readBoolean();
    int precision = in.readInt();
    PaillierPublicKey publicKey = new PaillierPublicKey(modulus);
    PaillierContext _context;
    _context = isSigned ? publicKey.createSignedContext(precision) : publicKey.createUnsignedContext(precision);
    contextF.set(this, _context);
    ciphertextF.set(this, _ciphertext);
    exponentF.set(this, _exponent);
    isSafeF.set(this, _isSafe);
    contextF.setAccessible(false);
    ciphertextF.setAccessible(false);
    exponentF.setAccessible(false);
    isSafeF.setAccessible(false);
  }
```

@wilko77 , @mhsiah , the above code would be more efficient although ugly, it is basically same to transfer necessary data. I tested it locally, it can pass the tests.

**Comment by mhsiah** [comment]
Made redundant by #57.

**Pull request closed without merging**